File: rinhaback2401/Logs.cs
using RinhaBack2401.Model;

namespace RinhaBack2401;

#if !EXTRAOPTIMIZE
public static partial class Logs
{
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "Got unhandled exception at url {url}:\n{exceptionMessage}.")]
    public static partial void AppError(this ILogger logger, string url, string exceptionMessage);

    [LoggerMessage(EventId = 2, Level = LogLevel.Trace, Message = "Inserted {rowsCount} rows into database. Details:\n{details}")]
    public static partial void DbInserted(this ILogger logger, int rowsCount, string? details);

    [LoggerMessage(EventId = 3, Level = LogLevel.Trace, Message = "Pool of {typeName} rented an item, has {itemsCount} before renting.")]
    public static partial void PoolRentingItem(this ILogger logger, string typeName, int itemsCount);

    [LoggerMessage(EventId = 4, Level = LogLevel.Trace, Message = "Pool of {typeName} returned an item, had {itemsCount} after return.")]
    public static partial void PoolReturnedItem(this ILogger logger, string typeName, int itemsCount);

    [LoggerMessage(EventId = 5, Level = LogLevel.Trace, Message = "Pool of {typeName} returning all items, had {itemsCount}.")]
    public static partial void PoolReturningAllItems(this ILogger logger, string typeName, int itemsCount);

    [LoggerMessage(EventId = 6, Level = LogLevel.Trace, Message = "Pool of {typeName} created with {itemsCount}.")]
    public static partial void PoolCreated(this ILogger logger, string typeName, int itemsCount);

    [LoggerMessage(EventId = 7, Level = LogLevel.Trace, Message = "Inserted transacao rows into database for client {idCliente}, tipo: {tipo}, valor: {valor}")]
    public static partial void DbInserted(this ILogger logger, int idCliente, int valor, TipoTransacao tipo);
}

public sealed class AppLogs { }
#endif

File: rinhaback2401vb/Logs.vb
Imports Microsoft.Extensions.Logging
Imports RinhaBack2401.Model
Imports System.Runtime.CompilerServices

#If Not EXTRAOPTIMIZE Then
Partial Public Module Logs
    <Extension()>
    Public Sub AppError(logger As ILogger, url As String, exceptionMessage As String)
        logger.LogError("Got unhandled exception at url {url}:" & vbLf & "{exceptionMessage}.", url, exceptionMessage)
    End Sub

    <Extension()>
    Public Sub DbInserted(logger As ILogger, rowsCount As Integer, details As String)
        logger.LogTrace("Inserted {rowsCount} rows into database. Details:" & vbLf & "{details}", rowsCount, details)
    End Sub

    <Extension()>
    Public Sub PoolRentingItem(logger As ILogger, typeName As String, itemsCount As Integer)
        logger.LogTrace("Pool of {typeName} rented an item, has {itemsCount} before renting.", typeName, itemsCount)
    End Sub

    <Extension()>
    Public Sub PoolReturnedItem(logger As ILogger, typeName As String, itemsCount As Integer)
        logger.LogTrace("Pool of {typeName} returned an item, had {itemsCount} after return.", typeName, itemsCount)
    End Sub

    <Extension()>
    Public Sub PoolReturningAllItems(logger As ILogger, typeName As String, itemsCount As Integer)
        logger.LogTrace("Pool of {typeName} returning all items, had {itemsCount}.", typeName, itemsCount)
    End Sub

    <Extension()>
    Public Sub PoolCreated(logger As ILogger, typeName As String, itemsCount As Integer)
        logger.LogTrace("Pool of {typeName} created with {itemsCount}.", typeName, itemsCount)
    End Sub

    <Extension()>
    Public Sub DbInserted(logger As ILogger, idCliente As Integer, valor As Integer, tipo As TipoTransacao)
        logger.LogTrace("Inserted transacao rows into database for client {idCliente}, tipo: {tipo}, valor: {valor}", idCliente, tipo, valor)
    End Sub
End Module

Public NotInheritable Class AppLogs
    End Class
#End If

